prow/cmd/generic-autobumper/main_test.go

```diff
@@ -0,0 +1,843 @@
+			upstreamURL:            "auto",
+			upstreamServerResponse: "     image: gcr.io/k8s-prow/deck:v20200717-cf288082e1",
+			expectedRes:            "v20200717-cf288082e1",
+			expectError:            false,
+			prefix:                 "gcr.io/k8s-prow/",
+		},
+		{
+			description:            "a valid response will return the correct tag with other prefixes in the same file",
+			upstreamURL:            "auto",
+			upstreamServerResponse: "other random garbage\n image: gcr.io/k8s-other/deck:v22222222-cf288082e1\n image: gcr.io/k8s-prow/deck:v20200717-cf288082e1\n     image: gcr.io/k8s-another/deck:v11111111-cf288082e1",
+			expectedRes:            "v20200717-cf288082e1",
+			expectError:            false,
+			prefix:                 "gcr.io/k8s-prow/",
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.description, func(t *testing.T) {
+			if tc.upstreamURL == "auto" {
+				// generate a test server so we can capture and inspect the request
+				testServer := httptest.NewServer(http.HandlerFunc(func(res http.ResponseWriter, req *http.Request) {
+					res.Write([]byte(tc.upstreamServerResponse))
+				}))
+				defer func() { testServer.Close() }()
+				tc.upstreamURL = testServer.URL
+			}
+
+			res, err := parseUpstreamImageVersion(tc.upstreamURL, tc.prefix)
+			if res != tc.expectedRes {
+				t.Errorf("The expected result %q != the actual result %q", tc.expectedRes, res)
+			}
+			if tc.expectError && err == nil {
+				t.Errorf("Expected to get an error but the result is nil")
+			}
+			if !tc.expectError && err != nil {
+				t.Errorf("Expected to not get an error but got one: %v", err)
+			}
+		})
+	}
+}
+
+func TestUpstreamImageVersionResolver(t *testing.T) {
+	prowProdFakeVersion := "v-prow-prod-version"
+	prowStagingFakeVersion := "v-prow-staging-version"
+	boskosProdFakeVersion := "v-boskos-prod-version"
+	boskosStagingFakeVersion := "v-boskos-staging-version"
+	prowRefConfigFile := "prow-prod"
+	boskosRefConfigFile := "boskos-prod"
+	prowStagingRefConfigFile := "prow-staging"
+	boskosStagingRefConfigFile := "boskos-staging"
+	fakeUpstreamURLBase := "test.com"
+	prowPrefix := "gcr.io/k8s-prow/"
+	boskosPrefix := "gcr.io/k8s-boskos/"
+	doesNotExistPrefix := "gcr.io/dne"
+	doesNotExist := "DNE"
+
+	prowPrefixStruct := Prefix{
+		Prefix:               prowPrefix,
+		RefConfigFile:        prowRefConfigFile,
+		StagingRefConfigFile: prowStagingRefConfigFile,
+	}
+	boskosPrefixStruct := Prefix{
+		Prefix:               boskosPrefix,
+		RefConfigFile:        boskosRefConfigFile,
+		StagingRefConfigFile: boskosStagingRefConfigFile,
+	}
+	//Prefix used to test when a tag does not exist. This is used to have parser return a tag that will make TagExists return false
+	tagDoesNotExistPrefix := Prefix{
+		Prefix:               doesNotExistPrefix,
+		RefConfigFile:        doesNotExist,
+		StagingRefConfigFile: doesNotExist,
+	}
+
+	fakeImageVersionParser := func(upstreamAddress, prefix string) (string, error) {
```

Comment:
Consider making this part of the table, which allows me to provide an image parser laser focused for the scenario, which tends to be easier to read.

It will be harder for me 3 months from now to try and figure out if it is safe for me to modify `case fakeUpstreamURLBase + "/" + prowRefConfigFile:` right now as every unit test depends on it.

So instead you can do this


```golang
cases := []struct {
  parser func(string, string) (string, error)
  currentTag string
}{
  {
      description: "whatever",
      currentTag: "cute-joke-to-make-this-more-fun",
      parser: func(upAddr, pref string) (string, error) {
         if upAddr == "cute-joke-to-make-this-more-fun" {
            return pref, nil
         }
         return "", errors.New("try again, you screwed up!")
    },
}
```

This tends to be easier to maintain -- easier to read, less brittle. Easier to read since I only need to care about the parser function inside the scope of the table case. Less brittle because I can add a new case and add a new parser for that case, rather than trying to figure out the best way to reuse and/or update the global parser to handle my additional scenario.